hw/pci-bridge/cxl_socket_transport.c

```diff
@@ -354,30 +354,51 @@ bool send_cxl_io_mem_write(int socket_fd, hwaddr hpa, uint64_t val, int size,
 
     assert(size % 4 == 0); // Ensure size is dword aligned
 
-    cxl_io_mem_wr_packet_t packet;
+    cxl_io_mem_wr_packet_32b_t packet;
+    cxl_io_mem_wr_packet_64b_t packet_64;
+    bool successful;
 
-    packet.system_header.payload_type = CXL_IO;
-    packet.system_header.payload_length = sizeof(packet);
+    uint16_t hdr_length = round_up_to_nearest_dword(size) / 4;
 
-    packet.cxl_io_header.fmt_type = MWR_64B;
-    uint16_t hdr_length = round_up_to_nearest_dword(size);
-    packet.cxl_io_header.length_upper = EXTRACT_UPPER_2(hdr_length);
-    packet.cxl_io_header.length_lower = EXTRACT_LOWER_8(hdr_length);
+    if (size > 4) {
+        assert(size == 8);
+        assert(val < 0xFFFFFFFFFFFFFFFF);
+        packet_64.system_header.payload_type = CXL_IO;
 
-    packet.mreq_header.req_id = 0;
-    packet.mreq_header.tag = *tag;
+        packet_64.cxl_io_header.fmt_type = MWR_64B;
 
-    hpa = htonll(hpa);
+        packet_64.cxl_io_header.length_upper = EXTRACT_UPPER_2(hdr_length);
+        packet_64.cxl_io_header.length_lower = EXTRACT_LOWER_8(hdr_length);
 
-    packet.mreq_header.addr_upper =
-        (hpa >> 8) & 0xFFFFFFFFFFFFFF; 
-    packet.mreq_header.addr_lower = (hpa & 0xFF) >> 2; 
+        packet_64.mreq_header.req_id = 0;
+        packet_64.mreq_header.tag = *tag;
 
-    packet.data = val;
+        packet_64.mreq_header.addr_lower = (hpa & 0xFF) >> 2;
+        packet_64.mreq_header.addr_upper = htonll(hpa) & 0xFFFFFFFFFFFFFF;
+        packet_64.system_header.payload_length = sizeof(packet_64);
+        packet_64.data = val;
+        successful = write(socket_fd, &packet_64, sizeof(packet_64)) != -1;
+        trace_cxl_socket_debug_num("MWR_64B Packet Size", sizeof(packet_64));
+    } else {
+        assert(size == 4);
+        assert(val < 0xFFFFFFFF);
+        packet.system_header.payload_type = CXL_IO;
 
-    trace_cxl_socket_debug_num("MRD_64B Packet Size", sizeof(packet));
+        packet.cxl_io_header.fmt_type = MWR_32B;
 
-    bool successful = write(socket_fd, &packet, sizeof(packet)) != -1;
+        packet.cxl_io_header.length_upper = EXTRACT_UPPER_2(hdr_length);
+        packet.cxl_io_header.length_lower = EXTRACT_LOWER_8(hdr_length);
+
+        packet.mreq_header.req_id = 0;
+        packet.mreq_header.tag = *tag;
+
+        packet.mreq_header.addr_lower = (hpa & 0xFF) >> 2;
+        packet.mreq_header.addr_upper = htonll(hpa) & 0xFFFFFFFFFFFFFF;
+        packet.system_header.payload_length = sizeof(packet);
+        packet.data = (uint32_t)(val & 0xFFFFFFFF);
+        successful = write(socket_fd, &packet, sizeof(packet)) != -1;
+        trace_cxl_socket_debug_num("MWR_32B Packet Size", sizeof(packet));
+    }
```

Comment:
Use a pointer to assign common fields, then use if/else-if statement to assign non-common fields.

Reply:
That was what I wanted to do, but `packet32` and `packet64` are of different types. A common pointer to assign values may cause issues. I still need to add extra code for type casting one type to the other, which will result in similar number of lines. This would be easier if the code is C++ (we can use templates then). 

I also tried memcpy `packet32` to `packet64` and change some values, but it would cause packet corruption.

Reply:
Temporarily cast it to a header-only type to assign the fields. e.g.
```
typedef struct {
    system_header_packet_t system_header;
    cxl_io_header_t cxl_io_header;
    cxl_io_mreq_header_t mreq_header;
} __attribute__((packed)) cxl_io_mem_packet_t;
```
I see 6 (or more) lines of code that could be shared.

Condition-based value determination followed by common "assignment code" could be useful here as well.